### DISCUSSION TITLE
Is it possible to type-check nested EloquentCollection?

### DISCUSSION THREAD

**Discussion by joelclermont** [larastan/larastan#1633] [Q&A]
Consider this method:

```php
    /**
     * @return EloquentCollection<string, EloquentCollection<int, Service>>
     */
    protected function getAvailableServicesByTypeForCompany(Company $company): Collection
    {
        return $company->availableServices()->get()->sortBy('description')->groupBy('type')->sortKeys();
    }
```

The `groupBy` creates a nested EloquentCollection where the outer collection is keyed by `type` (which is a `string`) and the inner collection is a typical Eloquent collection with int keys and model values.

But I get this error:

```
Type Illuminate\Database\Eloquent\Collection<int, App\Models\Service> in generic type
 Illuminate\Database\Eloquent\Collection<string, Illuminate\Database\Eloquent\Collection<int, App\Models\Service>> in PHPDoc tag @return is not subtype of template type TModel of
 Illuminate\Database\Eloquent\Model of class Illuminate\Database\Eloquent\Collection.
```

It seems that the type definition for EloquentCollection doesn't account for a nested collection. Is there a way to accomplish this?

**Comment by szepeviktor** [comment]
Our stub looks like this
https://github.com/nunomaduro/larastan/blob/ccac5b25949576807862cf32ba1fce1769c06c42/stubs/EloquentCollection.stub#L9-L10

**Reply by szepeviktor** [reply]
In outer EloquentCollection in place of TModel you have
`EloquentCollection<int, Service>`
which is not a model.

We have to resolve that.

**Reply by szepeviktor** [reply]
How about this?
```php
    /**
     * @return \Illuminate\Support\Collection<string, EloquentCollection<int, Service>>
     */
```

**Reply by joelclermont** [reply]
That signature results in this error

> Method App\Http\Controllers\Company\ServiceRateController::getAvailableServicesByTypeForCompany() should return Illuminate\Support\Collection<string, Illuminate\Database\Eloquent\Collection<int, App\Models\Service>> but returns Illuminate\Database\Eloquent\Collection<(int|string), Illuminate\Database\Eloquent\Collection<(int|string), App\Models\Service>>.
> Template type TValue on class Illuminate\Support\Collection is not covariant. Learn more: [What's Up With @template-covariant?](https://phpstan.org/blog/whats-up-with-template-covariant)

**Reply by joelclermont** [reply]
What if in the EloquentCollection stub it allowed TModel to either be Model|EloquentCollection?
